backend/src/order.ts

```diff
@@ -354,162 +336,252 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
+    successRedirect: config.public_url + 'api/orders/' + order_id + '/success',
+    errorRedirect: config.public_url + 'api/orders/' + order_id + '/failure',
+    successCallback: config.public_url + 'api/orders/' + order_id + '/notify/success',
+    errorCallback: config.public_url + 'api/orders/' + order_id + '/notify/failure',
+  };
+
+  log.log('debug', 'Creating payment with handler', {order_id});
+
+  [handler, err] = await to(getPaymentHandler(order_id));
+  if (err) {
+    return reject(err);
+  }
+
+  [createResp, err] = await to<ICreateResponse>(handler.create(order, args));
+  if (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    return reject(err);
+  }
+
+  log.log('debug','Update payment status', {order_id})
+  await db.query('update nk2_orders set payment_id = :payment_id, payment_url = :payment_url where id = :order_id', {payment_id: createResp.payment_id, payment_url: createResp.payment_url, order_id});
+  //Hack here a little bit, we are only interested in error, not result value
+  if (err) {
+    log.error('Failed to prepare payment', {error: err, order_id: order_id});
+    return reject(err);
+  }
+
+  log.log('debug', 'Updated payment info successfully', {order_id, res})
+
+
+  log.info('Created payment for order', {order_id});
+
+  return resolve(createResp);
 }
 
-export function paymentCancelled(order_id: number, params): Promise<any> {
-  log.info('Payment was cancelled', {order_id: order_id});
+export async function paymentDone(order_id: number, req: Request): Promise<any> {
 
-  var verification = [PAYTRAIL_PREFIX + order_id, params.TIMESTAMP, config.paytrail.password].join('|');
-  var verification_hash = md5(verification).toUpperCase();
+  let res, err: Error, successResp: ISuccessResponse, conn: IConnection, order: IOrder, handler: IPayment;
+  
+  log.log('debug', 'Create payment handler', {order_id});
+    [handler, err] = await to(getPaymentHandler(order_id));
+  if (err) {
+    return reject(err);
+  }
+
+  log.log('debug', 'Handle success request', {order_id});
+  [successResp, err] = await to(handler.handleSuccessCallback(req));
+  if (err) {
+    log.error('Payment succes handling failed', {error: err,order_id});
+    return reject(err);
+  }
+
+  log.info('Payment success handling succeeded', {order_id});
+
+
+  [res, err] = await to(updatePaymentStatusToPaid(order_id));
+  if (err) {
+    return reject(err);
+  }
 
-  if (verification_hash === params.RETURN_AUTHCODE) {
-    log.info('Verification hash matches - deleting order', {order_id: order_id});
-    return db.query('delete from nk2_orders where id = :order_id',
-      {order_id: order_id});
-  } else {
-    log.error('Hash verification failed!',
-      {order_id: order_id, verification_hash: verification_hash, return_authcode: params.RETURN_AUTHCODE});
-    throw 'Hash verification failed';
+  log.log('debug', 'Sending tickets', {order_id});
+  [res, err] = await to(sendTickets(order_id));
```

Comment:
This function can be called multiple times per order, so need to check here if the order was already paid or if the status was update (if you look at the previous code you'll see it chained the promises weirdly to do just that).

Reply:
Fixed, now payment status is first checked and`updated if required, tickets are send only if status was not yet paid.